File: archive/honey-crypto/src/threshold/sig.cc
#include "crypto/threshold/sig.hpp"
#include "blst/ops.hpp"
#include "threshold/interpolate.hpp"
#include "threshold/key_gen.hpp"
#include <cstdint>
#include <cstring>
#include <expected>
#include <span>
#include <string_view>
#include <system_error>
#include <vector>

namespace Honey::Crypto::Threshold::Sig {

namespace Constants {
    inline constexpr std::string_view DST_SIG = "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_";
} // namespace Constants

namespace {

    [[nodiscard]]
    auto combine_partial_signatures(
        const PublicParameters& public_params,
        std::span<const PartialSignature> partial_signatures)
        -> std::expected<Signature, std::error_code>
    {
        if (partial_signatures.size() != static_cast<size_t>(public_params.threshold)) {
            return std::unexpected(std::make_error_code(std::errc::invalid_argument));
        }

        return Crypto::Math::interpolate_at_zero(partial_signatures);
    }

    [[nodiscard]]
    auto verify_signature(const PublicParameters& params, BytesSpan message,
        const Signature& signature)
        -> std::expected<void, std::error_code>
    {

        auto err = bls::ops::core_verify_pk_in_g2(signature, params.master_public_key, true,
            message, as_span(Constants::DST_SIG), {});

        if (err) {
            return std::unexpected(err);
        }
        return {};
    }
}; // namespace
auto generate_keys(uint32_t players, uint32_t k)
    -> std::expected<KeySet, std::error_code>
{
    return Threshold::generate_keys<MasterPublicKey, VerificationKey>(players, k);
}

// 生成签名份额
[[nodiscard]]
PartialSignature sign(const PrivateKeyShare& share, BytesSpan message)
{
    auto h = bls::P1::from_hash(message, as_span(Constants::DST_SIG));

    bls::ops::sign_with(h, share.secret);

    return PartialSignature {
        .player_id = share.player_id,
        .value = h,
    };
}

// 验证单个签名份额
[[nodiscard]] auto verify_share(const PublicParameters& params,
    const PartialSignature& partial_signature,
    BytesSpan message)
    -> std::expected<void, std::error_code>
{
    const auto& partial_sig = partial_signature.value;
    int player_id = partial_signature.player_id;

    if (player_id < 1 || player_id > params.total_players) {
        return std::unexpected(std::make_error_code(std::errc::invalid_argument));
    }

    auto err = bls::ops::core_verify_pk_in_g2(
        partial_sig, params.verification_vector[player_id - 1], true, message,
        as_span(Constants::DST_SIG), {});
    if (err) {
        return std::unexpected(err);
    }
    return {}; // Success
}

[[nodiscard]]
auto validate(const PublicParameters& public_params, BytesSpan message,
    std::span<const PartialSignature> partial_signatures)
    -> std::expected<void, std::error_code>
{
    for (const auto& partial_sig : partial_signatures) {
        auto verify_result = verify_share(public_params, partial_sig, message);
        if (!verify_result) {
            return std::unexpected(verify_result.error());
        }
    }
    auto combined_sig_result = combine_partial_signatures(public_params, partial_signatures);
    if (!combined_sig_result) {
        return std::unexpected(combined_sig_result.error());
    }
    return verify_signature(public_params, message, *combined_sig_result);
}

} // namespace Honey::Crypto::Threshold::Sig

// ====================================================
// ==================== Unit Tests ====================
// ====================================================

#ifndef DOCTEST_CONFIG_DISABLE
#include <doctest/doctest.h>
#include <string>
#include <vector>

namespace Honey::Crypto::Threshold::Sig {

struct SigFixture {
    static constexpr int N = 10;
    static constexpr int K = 5;

    KeySet ks;

    SigFixture()
        : ks(*generate_keys(N, K))
    {
    }

    PublicParameters& params() { return ks.public_params; }
    std::vector<PrivateKeyShare>& shares() { return ks.private_shares; }
};

TEST_CASE("ThresholdSig.EndToEndFlow")
{
    constexpr int N = 10;
    constexpr int K = 5;

    auto result = generate_keys(N, K);
    REQUIRE(result.has_value());

    const auto& ks = *result;
    const auto& params = ks.public_params;
    const auto& shares = ks.private_shares;

    CHECK_EQ(params.total_players, N);
    CHECK_EQ(params.threshold, K);

    std::string msg = "HoneyBadger-GTest";

    std::vector<PartialSignature> partials;
    for (int id = 1; id <= K; ++id) {
        auto ps = sign(shares[id - 1], as_span(msg));

        auto vr = verify_share(params, ps, as_span(msg));
        CHECK(vr.has_value());

        partials.push_back(ps);
    }

    REQUIRE_EQ(partials.size(), K);

    auto final_verify = validate(params, as_span(msg), partials);
    CHECK(final_verify.has_value());
}

TEST_CASE("ThresholdSig.NotEnoughShares")
{
    constexpr int N = 10;
    constexpr int K = 5;

    auto result = generate_keys(N, K);
    REQUIRE(result.has_value());

    const auto& params = result->public_params;
    const auto& shares = result->private_shares;

    std::string msg = "FailTest";

    std::vector<PartialSignature> partials;
    for (int id = 1; id <= 2; ++id) {
        partials.push_back(sign(shares[id - 1], as_span(msg)));
    }

    auto res = validate(params, as_span(msg), partials);
    CHECK_FALSE(res.has_value());
}

TEST_CASE("ThresholdSig.InvalidShareVerification")
{
    constexpr int N = 10;
    constexpr int K = 5;

    auto result = generate_keys(N, K);
    REQUIRE(result.has_value());

    const auto& params = result->public_params;
    const auto& shares = result->private_shares;

    std::string msg = "CorrectMessage";
    std::string wrong = "WrongMessage";

    auto ps = sign(shares[0], as_span(msg));

    auto vr = verify_share(params, ps, as_span(wrong));
    CHECK_FALSE(vr.has_value());
}

TEST_CASE("ThresholdSig.DuplicatePlayerIds")
{
    constexpr int N = 5;
    constexpr int K = 3;

    auto result = generate_keys(N, K);
    REQUIRE(result.has_value());

    const auto& params = result->public_params;
    const auto& shares = result->private_shares;

    std::string msg = "DupTest";

    auto p1 = sign(shares[0], as_span(msg));
    auto p2 = sign(shares[0], as_span(msg)); // same player id
    auto p3 = sign(shares[1], as_span(msg));

    std::vector<PartialSignature> partials { p1, p2, p3 };

    auto res = validate(params, as_span(msg), partials);
    CHECK_FALSE(res.has_value());
}

TEST_CASE("ThresholdSig.InvalidPlayerId")
{
    constexpr int N = 5;
    constexpr int K = 3;

    auto result = generate_keys(N, K);
    REQUIRE(result.has_value());

    auto params = result->public_params;
    auto shares = result->private_shares;

    std::string msg = "BadId";

    auto ps = sign(shares[0], as_span(msg));
    ps.player_id = N + 1; // 非法 ID

    std::vector<PartialSignature> partials { ps };

    auto res = validate(params, as_span(msg), partials);
    CHECK_FALSE(res.has_value());
}

TEST_CASE_FIXTURE(SigFixture, "validate")
{
    const auto& params = SigFixture::params();
    auto& shares = SigFixture::shares();
    std::string msg = "HoneyBadger-GTest";

    SUBCASE("valid K shares")
    {
        std::vector<PartialSignature> partials;
        for (int id = 1; id <= K; ++id) {
            auto ps = sign(shares[id - 1], as_span(msg));
            CHECK(verify_share(params, ps, as_span(msg)));
            partials.push_back(ps);
        }

        CHECK(validate(params, as_span(msg), partials));
    }

    SUBCASE("not enough shares")
    {
        std::vector<PartialSignature> partials;
        for (int id = 1; id <= 2; ++id)
            partials.push_back(sign(shares[id - 1], as_span(msg)));

        CHECK_FALSE(validate(params, as_span(msg), partials));
    }

    SUBCASE("duplicate player ids")
    {
        auto p1 = sign(shares[0], as_span(msg));
        auto p2 = sign(shares[0], as_span(msg));
        auto p3 = sign(shares[1], as_span(msg));

        std::vector<PartialSignature> partials { p1, p2, p3 };
        CHECK_FALSE(validate(params, as_span(msg), partials));
    }

    SUBCASE("invalid player id")
    {
        auto ps = sign(shares[0], as_span(msg));
        ps.player_id = params.total_players + 1;

        CAPTURE(ps.player_id);
        CHECK_FALSE(validate(params, as_span(msg), std::span { &ps, 1 }));
    }

    SUBCASE("wrong message")
    {
        auto ps = sign(shares[0], as_span(msg));
        CHECK_FALSE(verify_share(params, ps, as_span("WrongMessage")));
    }
}

} // namespace Honey::Crypto::Threshold::Sig

#endif // DOCTEST_CONFIG_DISABLE

File: native/rust/src/crypto/threshold/sig.rs
use crate::crypto::crypto_error::CryptoError;

use crate::crypto::bls::{g1::G1, pairing::core_verify_pk_in_g2};

use super::keygen::{PartialSignature, SigPrivateKeyShare, SigPublicParams};

use crate::crypto::bls::interpolate::interpolate_at_zero;

const DST_SIG: &[u8] = b"BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_";

/// Produce a partial BLS signature from a key share.
pub fn sign(share: &SigPrivateKeyShare, msg: &[u8]) -> PartialSignature {
    let h = G1::hash_to_g1(msg, DST_SIG);
    let sig = h.scalar_mult(&share.secret);
    PartialSignature {
        player_id: share.player_id,
        value: sig,
    }
}

/// Verify a single partial signature against the player's verification key.
pub fn verify_share(
    params: &SigPublicParams,
    partial_sig: &PartialSignature,
    msg: &[u8],
) -> Result<(), CryptoError> {
    let id = partial_sig.player_id;
    if id < 1 || id > params.total_players {
        return Err(CryptoError::InvalidArgument(format!(
            "player_id {id} out of range 1..={}",
            params.total_players
        )));
    }
    let ver_key = &params.verification_vector[id - 1];
    if !core_verify_pk_in_g2(&partial_sig.value, ver_key, msg, DST_SIG) {
        return Err(CryptoError::VerificationFailed);
    }
    Ok(())
}

/// Combine k partial signatures via Lagrange interpolation and verify the result.
/// Returns the combined signature if verification succeeds.
pub fn combine_with_verify(
    params: &SigPublicParams,
    msg: &[u8],
    partial_sigs: &[PartialSignature],
) -> Result<G1, CryptoError> {
    // Verify each share first (also catches duplicate player_ids via verify_share checks)
    for ps in partial_sigs {
        verify_share(params, ps, msg)?;
    }

    if partial_sigs.len() != params.threshold {
        return Err(CryptoError::InsufficientShares {
            need: params.threshold,
            got: partial_sigs.len(),
        });
    }

    let share_pairs: Vec<(usize, G1)> = partial_sigs
        .iter()
        .map(|ps| (ps.player_id, ps.value))
        .collect();

    let combined = interpolate_at_zero(&share_pairs)?;

    if !core_verify_pk_in_g2(&combined, &params.master_public_key, msg, DST_SIG) {
        return Err(CryptoError::VerificationFailed);
    }
    Ok(combined)
}

#[cfg(test)]
mod tests {
    use super::super::keygen::{SigKeySet, generate_sig_keys};
    use super::*;

    struct Fixture {
        ks: SigKeySet,
        msg: Vec<u8>,
    }

    impl Fixture {
        fn new(n: usize, k: usize) -> Self {
            let ks = generate_sig_keys(n, k).unwrap();
            Fixture {
                ks,
                msg: b"HoneyBadger-GTest".to_vec(),
            }
        }
    }

    #[test]
    fn test_end_to_end() {
        let n = 10;
        let k = 5;
        let f = Fixture::new(n, k);
        let params = &f.ks.public_params;
        let shares = &f.ks.private_shares;

        let partials: Vec<_> = (0..k).map(|i| sign(&shares[i], &f.msg)).collect();

        for ps in &partials {
            verify_share(params, ps, &f.msg).unwrap();
        }

        combine_with_verify(params, &f.msg, &partials).unwrap();
    }

    #[test]
    fn test_not_enough_shares() {
        let f = Fixture::new(10, 5);
        let partials: Vec<_> = (0..2)
            .map(|i| sign(&f.ks.private_shares[i], &f.msg))
            .collect();
        assert!(combine_with_verify(&f.ks.public_params, &f.msg, &partials).is_err());
    }

    #[test]
    fn test_invalid_share_verification() {
        let f = Fixture::new(10, 5);
        let ps = sign(&f.ks.private_shares[0], &f.msg);
        let wrong_msg = b"WrongMessage";
        assert!(verify_share(&f.ks.public_params, &ps, wrong_msg).is_err());
    }

    #[test]
    fn test_duplicate_player_ids() {
        let f = Fixture::new(5, 3);
        let p1 = sign(&f.ks.private_shares[0], &f.msg);
        let p2 = sign(&f.ks.private_shares[0], &f.msg); // same player
        let p3 = sign(&f.ks.private_shares[1], &f.msg);
        let partials = vec![p1, p2, p3];
        assert!(combine_with_verify(&f.ks.public_params, &f.msg, &partials).is_err());
    }

    #[test]
    fn test_invalid_player_id() {
        let f = Fixture::new(5, 3);
        let mut ps = sign(&f.ks.private_shares[0], &f.msg);
        ps.player_id = 999;
        assert!(verify_share(&f.ks.public_params, &ps, &f.msg).is_err());
    }
}
